### PR TITLE
Maximum comments 

### PR DESCRIPTION
**Pull request by hajiler**

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -38,35 +38,42 @@
...
+    response.getWriter().println(gson.toJson(getDatastoreComments(maxComments)));
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String comment = request.getParameter("Comment:");
     long timeMillis = System.currentTimeMillis();
+    
+    //prevents adding empty comments to datastore
+    if (comment.length() != 0) {
```
Comment (jessrenteria) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Often with `String` and `Collection` in Java, using `!comment.isEmpty()` can be a little easier to read.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -38,35 +38,42 @@
...
-  public List<Comment> getDatastoreComments() {
+  public List<Comment> getDatastoreComments(int maxComments) {
     List<Comment> comments = new ArrayList<>();
 
     Query query = new Query("Comment").addSort("timeMillis", SortDirection.DESCENDING);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
+    int count = 0;
     for (Entity entity : results.asIterable()){
```
Comment (jessrenteria) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Consider using the overload of `asIterable` that takes in `FetchOptions`. You can limit the query to only fetch up to `maxComments` without needing the extra variable and bookkeeping.

**Review by jessrenteria — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (hajiler) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: So I did this, I also used forEach directly on the call instead of on the variable itself. But I'm on unsure if this is good practice for clean code because it is a pretty long sequence of function calls, but also the variable for the results isn't really needed.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -38,38 +39,39 @@
...
     Query query = new Query("Comment").addSort("timeMillis", SortDirection.DESCENDING);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    PreparedQuery results = datastore.prepare(query);
-
-    for (Entity entity : results.asIterable()){
-      comments.add(Comment.fromEntity(entity));
-    }
-
+    
+    datastore.prepare(query).asList(FetchOptions.Builder.withLimit(maxComments)).forEach((entity)-> {
```
Comment (jessrenteria) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:72]: People are kind of split on this.

On one hand, breaking out a ton of local variables might help call out the different types at play to reader.

On the other, this long sequence of functions calls returning an object which you call another function on is often referred to as "fluent" programming, and can often look really nice and readable. With what you have now, it's clear you're:

1. Starting with the Datastore API.
2. Preparing a query for it.
3. Fetching that query with a limit to a list.
4. Converting each comment and adding it to a list.

without needing to jump around and look at the types. If we go even further in this direction, you could make this whole function a one-liner:

```java
return DatastoreServiceFactory.getDatastoreService()
    .prepare(new Query("Comment").addSort("timeMillis", SortDirection.DESCENDING))
    .asList(FetchOptions.Builder.withLimit(maxComments))
    // Convert to a Java Stream [1].
    .stream()
    // Map the Entity objects to Comment objects, using your helper.
    // "Map" is a concept from functional programming which more or less means
    // "take a collection of type A, a function that take in an A and returns a B,
    // and return a collection of type B". In this case the "collection" is a Stream.
    // Note: Comment::entityToComment is called a method reference [2]. These looked scary the first time
    // I saw them because I didn't even know that syntax was possible...they can basically be used anywhere
    // where something from java.util.function is expected.
    .map(Comment::entityToComment)
    // Package up the Stream into a List (the list "collects" the elements of the stream).
    // java.util.stream.Collectors [3]
    .collect(Collectors.toList());
```

Some people would argue this becomes very hard to follow but IMO it's very readable (especially once you become comfortable with stream-based code). It kind of shifts the style of coding from imperative to declarative: rather than stating exactly what you want explicitly at a very low level, you start using higher and higher level abstractions (e.g., for loops / forEach is completely abstracted away with the higher-level map and collect concepts). Now it just looks like a pipeline of operations you're applying.

TOTALLY up to you if you want to go this far or keep it with what you have right now :)

[1] [Stream API](https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html)
[2] [Method references](https://docs.oracle.com/javase/tutorial/java/javaOO/methodreferences.html)
[3] [Collectors](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html)

**Review by jessrenteria — APPROVED** [review]
